src/Servers/Kestrel/Transport.MsQuic/Internal/QuicNativeMethods.cs

```diff
@@ -0,0 +1,484 @@
+            internal ConnectionEventDataStreamsAvailable StreamsAvailable;
+
+            [FieldOffset(0)]
+            internal ConnectionEventDataIdealSendBuffer IdealSendBuffer;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct ConnectionEvent
+        {
+            public QUIC_CONNECTION_EVENT Type;
+            internal ConnectionEventDataUnion Data;
+
+            public bool EarlyDataAccepted => Data.Connected.EarlyDataAccepted;
+            public ulong NumBytes => Data.IdealSendBuffer.NumBytes;
+            public IPEndPoint LocalAddress => null; // TODO
+            public IPEndPoint PeerAddress => null; // TODO
+            public QuicStream CreateNewStream(QuicRegistration registration)
+            {
+                return new QuicStream(registration, Data.NewStream.Stream, shouldOwnNativeObj: false);
+            }
+
+            public QUIC_STATUS ShutdownBeginStatus => Data.ShutdownBegin.Status;
+            public ushort ShutdownBeginPeerStatus => Data.ShutdownBeginPeer.ErrorCode;
+            public bool ShutdownTimedOut => Data.ShutdownComplete.TimedOut;
+            public ushort BiDirectionalCount => Data.StreamsAvailable.BiDirectionalCount;
+            public ushort UniDirectionalCount => Data.StreamsAvailable.UniDirectionalCount;
+            public QUIC_NEW_STREAM_FLAG StreamFlags => Data.NewStream.Flags;
+        }
+
+        [UnmanagedFunctionPointer(Api)]
+        internal delegate QUIC_STATUS ConnectionCallbackDelegate(
+         IntPtr Connection,
+         IntPtr Context,
+         ref ConnectionEvent Event);
+
+        internal delegate QUIC_STATUS ConnectionOpenDelegate(
+            IntPtr Session,
+            ConnectionCallbackDelegate Handler,
+            IntPtr Context,
+            out IntPtr Connection);
+
+        internal delegate QUIC_STATUS ConnectionCloseDelegate(
+            IntPtr Connection);
+
+        internal delegate QUIC_STATUS ConnectionStartDelegate(
+            IntPtr Connection,
+            ushort Family,
+            [MarshalAs(UnmanagedType.LPStr)]
+            string ServerName,
+            ushort ServerPort);
+
+        internal delegate QUIC_STATUS ConnectionShutdownDelegate(
+            IntPtr Connection,
+            uint Flags,
+            ushort ErrorCode);
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct StreamEventDataRecv
+        {
+            internal ulong AbsoluteOffset;
+            internal ulong TotalBufferLength;
+            internal QuicBuffer* Buffers;
+            internal uint BufferCount;
+            internal byte Flags;
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        internal struct StreamEventDataSendComplete
+        {
+            [FieldOffset(7)]
+            internal byte Canceled;
+            [FieldOffset(8)]
+            internal IntPtr ClientContext;
+
+            internal bool IsCanceled()
+            {
+                return !(0 == (1 & Canceled));
```

Comment:
Generally any non-zero value for a `BOOLEAN` is considered true. I'd suggest changing this to `return Canceled != 0;`.